Cargo RFC for min publish age (#3923)
rust-lang/rfcs: text/3923-cargo-min-publish-age.md

Reviewer: This revision (6b0ce8cc562912456af0a8d10fd13be8b1044f73) switched to `deny` by default and moved `fallback` to future possibilities based on [the conclusion](https://github.com/rust-lang/cargo-team/blob/main/meetings/sync-meeting/2026-04-07.md#min-publish-age) from the 2026-04-07 Cargo team meeting. In this revision, the behavior is more well-understood since it is the same as the pre-existing yanked versions mechanism. On the implementation side the change would also be small enough to ship it faster, and incrementally work on future extensions.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..6b0ce8cc562912456af0a8d10fd13be8b1044f73)*

Reviewer: @jlizen commented in <https://github.com/rust-lang/rfcs/pull/3923#issuecomment-4390038954>:

> One open question that isn't critical for implementation is, how to address the use case of:
> "I want to bypass my min publish age to build this new version of a crate that has a CVE fix, but I don't want to open the doors to the rest of the registry bypassing min age"
> 
> In other words, a shallow bypass, that lets you pull just one thing, if it has no also-new forced dependencies. Or if it does, you can add them one by one without opening the floodgates.

While not perfect, this is covered by `CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE=allow cargo update <pkg> --precise <version>`. It is scoped to a single invocation, just like how `cargo update <pkg> --precise <yanked-version>` works today. In the future, we would love to have something like cascading precise-update and exclusion list so you can pull in a sub-graph update without opening the entire door.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..f1244873dce8ced408286231d87bcbaa7cb98cc9)*

Reviewer 2: It would be nice if there were targeted informational messages when updating like this with the min publish age set and there being a newer version that suggests enabling the environment variable

> - Feature Name: cargo_min_publish_age

Reviewer 3: Honestly, in hindsight, this is definitely a bunch of concurrent arguments that I guess can be litigated separately, so, I will follow up on this in separate threads. In my head, it was a singular coherent argument of this being the wrong focus, but that honestly isn't a very constructive way to discuss this

- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

## Summary
[summary]: #summary

This proposal adds a new configuration option to cargo that specifies a minimum age for package
updates. When adding or updating a dependency, cargo won't use a version of that crate that
is newer than the minimum age, unless no possible version is older.

## Motivation
[motivation]: #motivation

There are a couple of reasons why one may wish not to use the most recent version of a package.

One such reason is to mitigate the risk of [supply chain attacks](https://en.wikipedia.org/wiki/Supply_chain_attack). Often, supply chain
attacks are found fairly quickly after they are deployed. Thus, if the dependency isn't updated
immediately after a release, you can have some protection against a new release of a dependency
containing malicious code. In light of recent supply chain attacks on the NPM ecosystem,
there has been an increased interest in using automated tools to ensure that packages used
are older than some age. This creates a window of time between when a dependency is compromised
and when that release is used by your project. See for example the blog post

> "[We should all be using dependency cooldowns](https://blog.yossarian.net/2025/11/21/We-should-all-be-using-dependency-cooldowns)".

Reviewer 4: > This really doesn't feel valid for one primary reason: people need to notice that there's a problem in order to mitigate it, and delaying mitigation also delays the time that people actually encounter an issue.

FWIW, the core thesis is that the relevant parties here are security scanners, not early victims. In the Python ecosystem, for example, the overwhelming majority of malicious package reports comes from automated static analysis, not from users who have already been victimized.

(And therefore it's not that _people_ need to notice a problem, it's that _systems_ need to be in place to monitor for problems and people should _generally_ wait for that monitoring rather than acting as ecosystem canaries themselves.)

Another reason to wish to delay using a new release, is because new versions can introduce new bugs. By only
using versions that have had some time to "mature", you can mitigate the risk of encountering those bugs a little.
Hopefully the bugs will be found before you start using the new version and thus could update to a version that fixes those bugs,
or lock your version, so you don't get the buggy version, if the bugs apply to you.

Although cargo, of course, allows you to specify exact versions in `Cargo.toml` and has a lock file that can freeze the versions used,
that requires manually inspecting each version of each transitive dependency to confirm it complies with a policy of
using a version older than some age.

As such, it would be useful to have an option to put a limit on commands like `cargo add` and `cargo update`
so that they can only use package releases that are older than some threshold.

> ## Guide-level explanation

Author: I've added a couple of examples.

[guide-level-explanation]: #guide-level-explanation

The `registry.global-min-publish-age` [configuration option][1][^1] for Cargo can be used to specify a minimum age for published versions to use.

When set, it contains a duration specified as an integer followed by a unit of "seconds", "minutes", "days", "weeks", or "months".
If a new crate would be added with a command such as `cargo add` or `cargo update`, it will use a version with a publish
time ("pubtime") before that is older than that duration, if possible.

The `resolver.incompatible-publish-age` configuration can also be used to control how `cargo` handles versions whose
publish time is newer than the min-publish-age. By default, it will try to use an older version, unless none is available
that also complies with the specified version constraint, or the `rust-version`. However by setting this to "allow"

> it is possible to disable the min-publish-age checking.

Reviewer: Added this to Unresolved Questions, though I think Mark has a point.

* The complexity might be unnecessary if we don't extend the resolver config beyond "allow" and "deny".
* Resolver behavior is not even a per-registry level config. If users want to permit a custom registry (which is trusted) but crates.io, they cannot use `resolver.incompatible-publish-age`. They still need to use per registry `CARGO_REGISTRY_<name>_MIN_PUBLISH_AGE=0` to turn it off. It is rare though valid as a use case.

I am slightly towards removing `resolver.incompatible-publish-age` even semantically it is appealing to me.

> it is possible to disable the min-publish-age checking.

Reviewer 5: Link to commit?

For myself, I'm unconvinced about removing it. While most users only have one registry, we need to still support the multi-registry users.  Uncommenting isn't accurate when doing things like running a "test latest versions" job in CI.  It is also very handy to have a simple, universally applicable example in our documentation.

This also keeps the way we talk about all of these things consistent.

> it is possible to disable the min-publish-age checking.

Reviewer: > Link to commit?

3696956f7b635c856c20f4a6d02fc8d717f5a938

> it is possible to disable the min-publish-age checking.

Reviewer: Removed in a2a748bdef618dfd9fb2725192212b224843907b

* Environment: `CARGO_REGISTRY_<name>_MIN_PUBLISH_AGE`

 Specifies the minimum timespan since a version's `pubtime` that it may be considered for `resolver.incompatible-publish-age` for packages from crates.io not set, `registry.global-min-publish-age` will be used.

 It supports the following values:

 * An integer followed by “seconds”, “minutes”, “hours”, “days”, “weeks”, or “months”
 * `"0"` to allow all packages

#### `registry.global-min-publish-age`

* Type: String
* Default: `"0"`
* Environment: `CARGO_GLOBAL_REGISTRY_<name>_MIN_PUBLISH_AGE`

 Specifies the global minimum timespan since a version's `pubtime` that it may be considered for `resolver.incompatible-publish-age` for packages. If `min-publish-age` is not set for a specific registry using `registries.<name>.min-publish-age`, Cargo will use this minimum publish age.

 It supports the following values:

* An integer followed by “seconds”, “minutes”, “hours”, “days”, “weeks”, or “months”
* `"0"` to allow all packages

### Behavior

In addition to what is specified above

* `min-publish-age` only apply to dependencies fetched from a registry, such as crates.io. They do not apply to git or path dependencies, in

>   part because there is not always an obvious publish time, or a way to find alternative versions.

Author: Done in 68641240f38130c404aa604634fb0301999ae9a1

* At this time, if a specific version is explicitly specified in Cargo.toml, or on the command line, that has higher precedence than the publish time check,
  and will be assumed to be valid. In the future it may be possible to change this behavior.
* `cargo add`
    * If a version is not explicitly specified by the user and the package is fetched from a registry (not a path or git), `min-publish-age` options
      will be respected.
* `cargo install`
    * If a specific version is not specified by the user, respect `registries.min-publish-age` for the version of the crate itself,

>       as well as transitive dependencies when possible.

Author: Personally, I think that having this apply to `cargo install` as a way of preventing installing an executable package that has been published very recently is also very useful. Although perhaps users may want a different policy for `cargo install` than for updating libraries?

There are also some significant differences between MSRV and this:

- For MSRV, the minimum rust version comes from the package Manifest, whereas for the minimum publish age, it comes from Carog configuration, which may be set by the user themselves. It seems odd to me that setting global-min-publish-age  in ~/.cargo/config.toml wouldn't apply to crates installed with `cargo install` and their dependencies.
- For MSRV and cargo install, generally the user doesn't care if crates are compatible with the minimum version in the package they are installing, so much as that they are compatible with the version of rust they are using to compile it?  Does the resolver at least prefer crates that are compatible with the version of rust that cargo is using?

That said, I do think it would be surprising for a .cargo/config.toml in the package itself to impact the resolver when running `cargo install`. 

So maybe we should specify that configuration from $CARGO_HOME/config.toml is used for checking minimum publish age in `cargo install` but not configuration from the package being installed itself?

>       as well as transitive dependencies when possible.

Reviewer 5: > Personally, I think that having this apply to cargo install as a way of preventing installing an executable package that has been published very recently is also very useful. Although perhaps users may want a different policy for cargo install than for updating libraries?

We've already started down the path of documenting `resolver` as not applying to `cargo install` and I think it would be a problem to be inconsistent on that.

We can separately look into an `install.resolver` table.

Or we could look into more general fixes for `cargo install` so that the installed package is also checksumed, like installing dependencies.

> That said, I do think it would be surprising for a .cargo/config.toml in the package itself to impact the resolver when running cargo install.

`cargo install` re-reloads the config from only `CARGO_HOME`

>       as well as transitive dependencies when possible.

Reviewer: I removed the mention entirely in https://github.com/rust-lang/rfcs/pull/3923/changes/64f02f5601a2245c0bd006bc938c9f203c3dfa0c. `cargo install` is not relevant anymore as pubtime-incompatible versions is just another yanked versions.

>       as well as transitive dependencies when possible.

Reviewer 5: Don't we still need to talk about `cargo install` because `Cargo.lock` is not used by default?

    * If the version of a crate in the lockfile is already newer than `min-publish-age`, then `cargo update` will not update that crate, nor will
      it downgrade to an older version. It will leave the version as it is.
* When a lockfile is generated, as with `cargo generate-lockfile` or other commands such as `cargo build` that can do so, then versions will be
  selected that comply with the `min-publish-age` policy, if possible.
* If the only version of a crate that satisfies the `min-publish-age` constraint is a yanked version, it will behave as if no versions satisfied the
  `min-publish-age` constraint. In other words, yanked versions has higher priority than the `min-publish-age` configuration.

## Drawbacks
[drawbacks]: #drawbacks

The biggest drawback is that if this is widely used, it could potentially lead to it taking longer for problems to be discovered after a version is published.
However, most likely, there will be a spread of values used, depending on risk tolerance, and hopefully the result is actually that there will be a more gradual rollout in most
cases.

## Rationale and Alternatives

> [rationale-and-alternatives]: #rationale-and-alternatives

Reviewer: > we should talk about the reason we are going with fallback and not deny is as a way to allow people to override this when there is a fix or feature they need. We've been experimenting with ways of doing this with cargo update --precise but if a newer transitive dependency is needed, you can't do that.

Switched to `deny` by default and moved `fallback` to future possibilities based on [the conclusion](https://github.com/rust-lang/cargo-team/blob/main/meetings/sync-meeting/2026-04-07.md#min-publish-age) from the 2026-04-07 Cargo team meeting. Not sure if this review comment is still relevant.

### Why not leave this to third party tools?

There are already some third party tools that fulfill this functionality to some degree. For example, dependabot and renovate can
be used for updating Cargo.toml and Cargo.lock, and both support some form of minimum publish age. And the cargo-cooldown project provides
an alternative to `cargo update` that respects a minimum publish age.

However, these tools only work for updating and adding dependencies outside of cargo itself, they do not
have any impact on explicitly run built-in cargo commands such as `cargo update` and `cargo add`.

> Having built-in support makes it easier to enforce a minimum publish age policy.

Author: I think that is only true if you don't have a Cargo.lock file, or your Cargo.lock file is missing something from your Cargo.toml file.

But I did try to mention that in https://github.com/rust-lang/rfcs/pull/3923/changes/d78605cb819480d64ef8a401250f2c22c5273995

  The current proposal is based on precedent of "credential-provider" and "global-credential-provider", but perhaps we shouldn't follow that precedent?
* How do we make it clear when things are held back?
    * The "locking" message for [Cargo time machine (generate lock files based on old registry state) #5221](https://github.com/rust-lang/cargo/issues/5221) lists one time but the time here is dependent on where any given package is from
    * We list MSRVs for unselected packages, should we also list publish times? I'm assuming that should be in local time
    * Locking message for [Cargo time machine (generate lock files based on old registry state) #5221](https://github.com/rust-lang/cargo/issues/5221) is in UTC time, see [Tracking Issue for _lockfile-publish-time_ #16271](https://github.com/rust-lang/cargo/issues/16271), when relative time differences likely make local time more relevant
* Implementation wise, will there be much complexity in getting per registry information into `VersionPreferences` and using it?
* `fallback` precedence between this and `incompatible-rust-version`?

> 

Author: Any such guarantee would need to be predicated on some conditions, including:

- The build was possible before the malicious release on the target platform and compiler version. I.E. there are no cases that require a fallback (a "deny" policy could potentially prevent a compromised build if that isn't the case, but that is currently put off as a future possibility)
- The malicious release happened more recently than the minimum publish age. 
- Either none of the versions needed for a secure build have been yanked OR there a lock file is used.

Hopefully the bugs will be found before you start using the new version and thus could update to a version that fixes those bugs,
or lock your version, so you don't get the buggy version, if the bugs apply to you.

Although cargo, of course, allows you to specify exact versions in `Cargo.toml` and has a lock file that can freeze the versions used,
that requires manually inspecting each version of each transitive dependency to confirm it complies with a policy of
using a version older than some age.

As such, it would be useful to have an option to put a limit on commands like `cargo add` and `cargo update`
so that they can only use package releases that are older than some threshold.

Note that this is **not** a full solution to compromised dependencies. It can increase the protection against certain types of
"supply chain" attacks, but not all of them. As such, using this feature should not be relied upon for security by itself.

## Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The `registry.global-min-publish-age` [configuration option][1][^1] for Cargo can be used to specify a minimum age for published versions to use.

> When set, it contains a duration specified as an integer followed by a unit of "seconds", "minutes", "days", "weeks", or "months".

Reviewer 5: We support `months` for GC already, see https://doc.rust-lang.org/cargo/reference/config.html#cacheauto-clean-frequency

I would prefer consistency

> When set, it contains a duration specified as an integer followed by a unit of "seconds", "minutes", "days", "weeks", or "months".

Reviewer 6: Sigh. Fair enough.

I do think "months" is a more reasonable configuration for GC than for this, but nonetheless, valid that this should favor consistency.

Let's at least ensure our documentation steers users towards sensible values.

- Feature Name: cargo_min_publish_age
- Start Date: 2026-02-23
- RFC PR: [rust-lang/rfcs#3923](https://github.com/rust-lang/rfcs/pull/3923)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

## Summary
[summary]: #summary

This proposal adds a new configuration option to cargo that specifies a minimum age for package
updates. When adding or updating a dependency, cargo won't use a version of that crate that

> is newer than the minimum age by default with a way to override to get urgent security fixes.

Reviewer 5: ```suggestion
This proposal adds a new configuration option to cargo that specifies a minimum age for package
updates. When adding or updating a dependency, cargo won't use a version of that crate that
is newer than the minimum age when specified, with a way to override to get urgent security fixes.
```

Wanting to make this clear that this feature is opt-in.  The "by default" was referring to when when specified and being overridenable but someone scanning could easily miss that detail

immediately after a release, you can have some protection against a new release of a dependency
containing malicious code. In light of recent supply chain attacks on the NPM ecosystem,
there has been an increased interest in using automated tools to ensure that packages used
are older than some age. This creates a window of time between when a dependency is compromised
and when that release is used by your project. See for example the blog post
"[We should all be using dependency cooldowns](https://blog.yossarian.net/2025/11/21/We-should-all-be-using-dependency-cooldowns)".

Another reason to wish to delay using a new release, is because new versions can introduce new bugs. By only
using versions that have had some time to "mature", you can mitigate the risk of encountering those bugs a little.
Different people (or groups of people) have different tolerance for risk, and this provides
a mechanism whereby new versions can roll out gradually to users depending on the tolerance
for risk of those users.

As such, it would be useful to have an option to put a limit on commands like `cargo add` and `cargo update`

> so that they can only use package releases that are older than some threshold.

Reviewer 5: Concrete suggestion for https://github.com/rust-lang/rfcs/pull/3923/changes#r2841902940
```suggestion
As such, it would be useful to have an option to put a limit on commands that update the `Cargo.lock` file (not just  `cargo add` and `cargo update` but other commands after editing `Carg.toml` like `cargo check`, etc)
so that they can only use package releases that are older than some threshold.
```

Note that this is **not** a full solution to compromised dependencies. It can increase the protection against certain types of
"supply chain" attacks, but not all of them. As such, using this feature should not be relied upon for security by itself.

## Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The `registry.global-min-publish-age` [configuration option][1][^1] for Cargo can be used to specify a minimum age for published versions to use.

When set, it contains a duration specified as an integer followed by a unit of "seconds", "minutes", "days", or "weeks".

> If a new crate would be added with a command such as `cargo add` or `cargo update`, it will use a version with a publish

Reviewer 7: “If a new crate would be added” is confusing. Taken colloquially, it sounds like min-publish-age would only apply to new packages (i.e. `cargo add` or updating adding a new dependency) and not `cargo update`s to existing packages; but I suspect it actually means new *versions* of packages whether or not that package name was previously in the dependency graph.

Reviewer 5: This wording was removed in 0689bbf55d49918a8f12bc6c2207efc3faca1fad

it is possible to disable the min-publish-age checking.

Most likely, `resolver.incompatible-publish-age` will usually be left at its default of `fallback`, however it may occasionally
be desirable to use it to temporarily turn off the minimum age check, especially if there are configurations for multiple
registries.

It is also possible to configure the `min-publish-age` per cargo registry. `registries.<name>.min-publish-age` sets
the minimum publish age for the `<name>` registry. And `registry.min-publish-age` sets it for the default registry
crates.io registry.

### Using newer version

In some cases, it may be desirable to use a version that is newer than the minimum publish age. For example, because a new
version has a critical security fix, or because it is part of the same family of crates as the dependent crate, and they should
be released together.

If `resolver.incompatible-publish-age` is "fallback" (the default), it is possible to bypass the check by updating the version range to require
the newer version in `Cargo.toml`, or with `cargo add`, or specify the exact version to use with `cargo update --precise`.

> In the future, additional controls may be provided (see [Future Possibilities](#future-possibilities)).

Author: I agree that could be useful. I'm not sure if that belongs as part of this RFC though.

> In the future, additional controls may be provided (see [Future Possibilities](#future-possibilities)).

Reviewer 5: I agree that this is independent though would be good to mention it in the Future Possibilities if we don't already.

> In the future, additional controls may be provided (see [Future Possibilities](#future-possibilities)).

Reviewer: Added in https://github.com/rust-lang/rfcs/pull/3923/changes/5a47fdf892f31c76f2c9db234ac416b88928d1e7

     * Values:

       * allow: treat rust-version-incompatible versions like any other version
       * fallback: only consider rust-version-incompatible versions if no other version matched

 [`package.resolver`](https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions) is only a version number. When adding `incompatible-rust-version`, we intentionally deferred anything being done in manifests.

 [`[registry]`](https://doc.rust-lang.org/cargo/reference/config.html#registry)

     * Set default registry

     * Sets credential providers for all registries

     * Sets crates.io values

 [`[registries]`](https://doc.rust-lang.org/cargo/reference/config.html#registries)

     * Sets registry specific values

 `yanked`: can't do new resolves to it but left in if already there. Unstable support to force it with `--precise` but that doesn't apply recursively.

 pre-release: requires opt-in through version requirement. Unstable support to force it with `--precise` but that doesn't apply recursively.

 We use the term `publish` and not `release`

> ## Unresolved Questions

Reviewer: > do we still need publish if its in the registry table? Probably yes to make it clear what is looked at and in shorthand communication.

Yes otherwise it could also be yank time (though crates.io doesn;t expose that yet)

> should we leave room for adding absolute times?

The only use case I can imagine is <https://github.com/rust-lang/cargo/issues/5221>, which we have the unstable `cargo generate-lockfile --publish-time` to provide a time machine. However, `cargo generate-lockfile --publish-time` is not perfect because it updates the entire lockfile, and people may have some deps wanted to be pinned.

If this doesn't have room for absolute times, we might need something like `cargo update --publish-time` to help people selectively upgrade/downgrade specific packages. IMHO that would be a bad idea. I'd like to see all CLI flag can be specified in Cargo config rather than one-off flag in only CLI.

    * Locking message for [Cargo time machine (generate lock files based on old registry state) #5221](https://github.com/rust-lang/cargo/issues/5221) is in UTC time, see [Tracking Issue for _lockfile-publish-time_ #16271](https://github.com/rust-lang/cargo/issues/16271), when relative time differences likely make local time more relevant
* Implementation wise, will there be much complexity in getting per registry information into `VersionPreferences` and using it?
* `fallback` precedence between this and `incompatible-rust-version`?
* Can we, and should we make any guarantees about security when using this feature, such as "a release of a malicious version of a crate will not compromise the build

## Future Possibilities
[future-possibilities]: #future-possibilities

- Support "deny" for `resolver.incompatible-publish-age`.
    - This is initially excluded, because it isn't clear how this should behave with respect to versions already in Cargo.lock, or use with the `--precise` flag.
    - What would an error look like?
    - How would you be able to override this for specific crates for important security updates, or for related crates that should be released at the same time?

> - Add a way to specify that the minimum age doesn't apply to certain packages. For example, by having an array of crates that should always use the newest version.

Reviewer: Moved in https://github.com/rust-lang/rfcs/pull/3923/changes/bcbcb7593ce9ac335d97cb13c89b2c36b9be4828

You can pin versions in your `Cargo.toml` but that is a manual process and doesn't cover transitive
dependencies.

`Cargo.lock` records versions but those are at the time of last change.
Adding a new dependency can cause you to pull in transitive dependencies that are outside
your desired minimum age. There isn't a manageable way to run `cargo update` and intentionally
get versions that are inside of your desired minimum age.

### Configuration Locations and Names

The locations and names of the configuration options in this proposal were chosen to be
consistent with existing Cargo options, as described in [Related Options in Cargo](#related-options).

The term "publish" was used rather than "package", "version", or "release" to make it
clear that this only applies to crates that are published in a registry.

### fallback and deny

We default `resolver.incompatible-publish-age` to "fallback" rather than deny
and defer support for "deny" to future possibilities, because that allows user to allow
users to easily override the minimum age for specific crates when necessary.

Specifically, with "fallback" it is possible to override the minimum age behavior for

> specific crates by specifying a more specific version in Cargo.toml, or using `cargo update --precise`.

Reviewer 8: FTR, with such a default choice this feature would only be minimally useful for security.

I can have `foobar = "0.1"` in my Cargo.toml. Currently, there are foobar `0.1.1,0.1.2,0.1.3` releases, I am locked to `0.1.2`.

Then a hacker gets hold of `foobar` publishing credentials and publishes foobar `0.1.4`, yanking all of the existing foobar versions. Cargo starts warning that the dependency has been yanked, and suggests you to use a newer version. A `cargo update` later, and you are compromised. Worse if you are doing `cargo install`.

I say minimally useful because `fallback` does protect you from hacked publish-only credentials.

A good solution includes both having an excludes array that you manually vet and maintain for security updates, as well as defaulting to `deny`.

Author: hmm, that's a fair point that an attacker might yank older, stable versions.

One possible way that could be mitigated is that if:

- The version in the Cargo.lock file is yanked
- There is no non-yanked version newer than the current version in the Cargo.lock and older than the minimum age
- And there is a version newer than the minimum age that isn't yanked

Then cargo will prefer the  yanked version that has already been selected in Cargo.lock over the new version that isn't yanked.

That still has a weakness if there isn't a Cargo.lock file though. 

I'm not sure that even if we supported `deny` that we would need  an exclude list, as long as versions that are already in Cargo.lock are considered safe, even if they are newer than the minimum age and the policy is deny.  Or said another way, the Cargo.lock file could itself act as an excludes list.

Reviewer 5: > A good solution includes both having an excludes array that you manually vet and maintain for security updates, as well as defaulting to deny.

I consider an exclude list to be a similarly bad solution as there is no way to know when to remove items from it and that then opens you up to attack.

Reviewer 8: > Then cargo will prefer the yanked version that has already been selected in Cargo.lock over the new version that isn't yanked.

I'm not sure this is easy to represent in the resolver algorithm, there might be edge cases. I'd say it's easier to reason about if we just treat all crates newer than the configured exclusion period as if they were yanked.

> I consider an exclude list to be a similarly bad solution as there is no way to know when to remove items from it and that then opens you up to attack.

looking at the pnpm exclusion design, you apparently put package names there, applying to every version of the package. Where yes, it's harder to remove items as you don't have a good criterion when to remove them.

Most crate authors publish only one security fix release anyway, I've seen it rarely that they also publish security fixes for older semver incompatible branches.

If we require the full version to be spelled out, for example with [replace section](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-replace-section) syntax, then cargo can start emitting warnings the moment those versions are older than the exclusion period.

Reviewer 5: > If we require the full version to be spelled out, for example with [replace section](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-replace-section) syntax, then cargo can start emitting warnings the moment those versions are older than the exclusion period.

If we went with [PackageIdSpec syntax](https://doc.rust-lang.org/cargo/reference/pkgid-spec.html), I'd want this to start with
- error for git, path sources
  - Not relevant
- error for name-only syntax
  - avoid people under-specifying and getting too many updates
- error if the version is incomplete
  - avoid people under-specifying and getting too many updates
  - only registry sources are supported so version is always relevant

An annoyance with an exclude list and `deny` would be transitive dependencies.  You will need to exclude every single one where the version requirement asks for too-new of a version.  Say you do this with `clap`.  That is really `clap`, `clap_derive`, `clap_builder`, and `clap_lex`.  We can't make exceptions because that would prevent the backtracking that would be expected with a `deny` mode.  At least if we had #3826, we would drop the coupled proc-macro / macro-exporter package split.

If we go with `deny`, we'd probably always respect the lockfile even if it has too-new packages, like with `yank`.

There may also be a desire to still support `cargo update --precise <too-new>`.  We've been experimenting with [`precise-pre-release`](https://doc.rust-lang.org/cargo/reference/unstable.html#precise-pre-release) and talked about extending that to yanked as well.  The problem with that approach is it is shallow: you can't "bless" a transitive yanked dependency at the same time.  That can be even more of a problem for this "too new" case because they are likelier to have "too new" deps (e.g. the `clap` example) while an entire dependency tree doesn't normally get yanked.  I've been considering proposing a `fallback` mode for yanked packages because of this.   At least if we have both `fallback` and `deny`, users can temporarily switch to change their lockfile.

Reviewer 5: To be clear, if we `deny` on this and if something is both yanked and too new, then `cargo generate-lockfile` will fail.

Reviewer 5: As what we do with fallback vs deny can have a large impact on the design and text of the RFC, I took this to the Cargo team to get a vibe check so more thorough reviews can be on what would hopefully be closer to done.

We are sympathetic to the problems with `fallback` and yanking.  We lean towards going ahead with a `deny` mode but treating it like yanked
- `Cargo.lock` is always respected
- `cargo update --precise` is always respected (I forgot we stabilized this for yanked)

The way to allow something would be either `cargo update --precise` or `CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE=allow`. If the new version is that important, users should consider updating their version requirement as well so that the desire for the new version is recorded more permanently than what we have with configuration.  This means that the resolve would fail if `cargo generate-lockfile` is run / the lockfile is deleted.

With this, we are aiming for a minimal set of features to stabilize that can fulfill the different use cases.  A `fallback` mode, an allow list, specifying precise on multiple packages, etc can always be added in the future based on what we learn about use cases.

Caveats
- The resolver can error in some cases because it is being too strict about updating of dependencies.  We likely need to evaluate and improve
- We would like to improve how the status of packages is reported but don't have suggestions at the moment on how that would look

Reviewer 8: > To be clear, if we `deny` on this and if something is both yanked and too new, then `cargo generate-lockfile` will fail.

Yeah I see this as a feature, not a bug. There are three alternatives here really:

* resolve to the yanked version -> seems kinda unexpected to me, if this is desired, we should provide an explicit `--ignore-yanked` flag or similar (cc https://github.com/rust-lang/cargo/issues/4225)
* resolve to the too new version -> doesn't provide security benefits (explained above), the attacker might have yanked the older versions purposefully to activate the fallback
* error -> attacker can still DOS the CIs of dependent projects but that's the smallest evil, at least no compromise happens (and downstream projects can add lockfiles)

Reviewer: The latest revision (6b0ce8cc562912456af0a8d10fd13be8b1044f73) switched to `deny` by default and moved `fallback` to future possibilities based on [the conclusion](https://github.com/rust-lang/cargo-team/blob/main/meetings/sync-meeting/2026-04-07.md#min-publish-age) from the 2026-04-07 Cargo team meeting.

 [`package.resolver`](https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions) is only a version number. When adding `incompatible-rust-version`, we intentionally deferred anything being done in manifests.

 [`[registry]`](https://doc.rust-lang.org/cargo/reference/config.html#registry)

     * Set default registry

     * Sets credential providers for all registries

     * Sets crates.io values

 [`[registries]`](https://doc.rust-lang.org/cargo/reference/config.html#registries)

     * Sets registry specific values

 `yanked`: can't do new resolves to it but left in if already there. Unstable support to force it with `--precise` but that doesn't apply recursively.

 pre-release: requires opt-in through version requirement. Unstable support to force it with `--precise` but that doesn't apply recursively.

 We use the term `publish` and not `release`

## Unresolved Questions
[unresolved-questions]: #unresolved-questions

* Would it be better to have `registry.min-publish-age` be the global setting, and `registries.crates-io.min-publish-age` be the setting for the crates.io registry?

>   The current proposal is based on precedent of "credential-provider" and "global-credential-provider", but perhaps we shouldn't follow that precedent?

Reviewer 9: Reading the RFC I found myself wondering what "global-" meant here, since in the rest of [Cargo config docs](https://doc.rust-lang.org/cargo/reference/config.html) "global" refers to user settings ("global cache", "global git configuration", "globally change profile settings"). So I don't think the precedent is great?

Reviewer 5: Unfortunately, there is already precedence that `[registry]` is for crates.io and we have already set the precedence that to use a `global-` prefix to disambiguate it.

Updating index
 Locking 1 package to recent Rust 1.60 compatible version
  Adding some-package v1.2.3 (available: v1.6.0, published 2 days ago)
```

While a CI job runs:
```
env:
  CARGO_RESOLVER_INCOMPATIBLE_RUST_VERSIONS: allow
  CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE: allow
steps:
  - uses: actions/checkout@v4
  - run: rustup update stable && rustup default stable
  - run: cargo update --verbose
  - run: cargo build --verbose
  - run: cargo test --verbose
```

This will mean that:
- Locally, `cargo add foo` will default the version requirement on `foo` to be low enough to support the 7 day old package
- Locally, `cargo update` will update your `Cargo.lock` to versions within the minimum-release age
- This CI job will verify the latest versions of your dependencies

Note: this check does not apply to
- path dependencies
- git dependencies
- registries that do not include `pubtime` (crates.io supports it)

### Per-registry configuration

It is also possible to configure the `min-publish-age` per cargo registry.
`registries.<name>.min-publish-age` sets the minimum publish age for the `<name>` registry.
And `registry.min-publish-age` sets it for crates.io.

For example:
```toml

> [registry.my-org]

Reviewer: [`[registries]`](https://doc.rust-lang.org/cargo/reference/config.html#registries) table you meant I believe
 
```suggestion
[registries.my-org]
```

index = "https://my.org"
min-publish-age = 0 # this registry is fully trusted

[registry]
# Default for any registry without a specific value
global-min-publish-age = "14 days"

> # Value to use for crates.io

Reviewer: teeny tiny nit (and through out this RFC):

```suggestion
# Value to use for the default registry
```

Granted, in Cargo doc we also assume the default is crates.io, and we probably don't want to confuse people for just being more comprehensive.

Reviewer 5: That is how the parallel docs for credential provider discuss it

And `registry.min-publish-age` sets it for crates.io.

For example:
```toml
[registries.my-org]
index = "https://my.org"
min-publish-age = "0" # this registry is fully trusted

[registry]
# Default for any registry without a specific value
global-min-publish-age = "14 days"
# Value to use for crates.io
min-publish-age = "5 days"
```

This will use a minimum publish age of
- 5 days for crates.io
- no minimum for `my-org`
- 14 days for any other registry.

### When no version matches

If no version of a dependency satisfies both the version requirement and the minimum publish age,
the resolve will error, similar to when all matching versions are yanked:

```console
$ cargo update
error: failed to select a version for the requirement `some-package = "^1.3"`
  version 1.3.0 is too new (published 2 days ago, minimum age 14 days)
```

### Using newer versions

In some cases, it may be desirable to use a version that is newer than the minimum publish age.
For example, `some-package` from [earlier](#guide-level-explanation) has a fix for a vulnerability in v1.3.0.

Since too-new versions follow yanked semantics,
the same override mechanisms apply:

```console

> $ cargo update some-package --precise 1.3.0

Reviewer 5: imo `--precise` is too broken that I wouldn't put a focus on it in this document.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..fcbe967ea774398cb47185f7aed66e48905ccb2a)*

Reviewer: De-emphasize a bit in 3bad7c983d789deb275ab48ee323e6e53339665e

It is still worth keeping around since it is the current stable way to deal with the stuck situation.

warning: selected package `some-package@1.3.0` is too new
  = note: published 2 days ago, minimum age 14 days
    Updating some-package v1.2.3 -> v1.3.0

> ```

Reviewer 5: This doesn't fit within the current style of how we warn about these types of things and doesn't scale up.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..fcbe967ea774398cb47185f7aed66e48905ccb2a)*

Reviewer: We have [a similar warning](https://github.com/rust-lang/cargo/blob/3c4678d6a93a072f0a6cf379ea673e35abbe415f/tests/testsuite/update.rs?plain=1#L1478-L1486
) for yanked (yeah I missed the vertical bar `|`):

```
[UPDATING] `dummy-registry` index
[WARNING] selected package `bar@0.1.1` was yanked by the author
  |
  = [HELP] if possible, try a compatible non-yanked version
[UPDATING] bar v0.1.0 -> v0.1.1
```

* Which part of it doesn't scale up?
* should we consolidate published and minimum age in the warning title?

Reviewer 5: 1. I would lean more towards reporting this the same way we report MSRV
2. This is reminding me how much of a mess yanked is and how much worse precise yanked made it.  I think there was a comment I forgot to respond to about "like yanked but different" being smaller but I actually think it could make it take more work because I am very hesitant about reproducing how we do things with yanked.

Reviewer: Just FYI, this warning example will be removed as part of de-emphasizing `--precise`.

Reviewer: Removed in 3f26931afe44ecfea4ff0c5595e26c22c55c771d

To record the intent more permanently,
bump the version requirement in `Cargo.toml`.

For a broader override, the `CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE=allow` environment variable

> disables the check entirely.

Reviewer 5: I feel like this may be misconstrued as "or you can completely disable it".

Could we just put in an example:

````suggestion
A limitation of `cargo update --precise` is that you can only force the use of that one package but it may have dependencies on new versions as well.
You can temporarily disable the check to accomplish this:
```console
$ CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE=allow cargo update clap --precise 4.5.3
Updating clap 4.3.0 -> 4.5.3 (published 2 days ago, minimum age 14 days)
Updating clap_derive 4.3.0 -> 4.5.3 (published 2 days ago, minimum age 14 days)
Updating clap_builder 4.3. -> 4.5.3 (published 2 days ago, minimum age 14 days)
```
For a broader override, the `CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE=allow` environment variable
disables the check entirely.
````

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..fcbe967ea774398cb47185f7aed66e48905ccb2a)*

Reviewer: Added in. 749164f2e6ba345046585b41d0b3522bebf98e66

[1]: https://doc.rust-lang.org/cargo/reference/config.html
[^1]: As specified in `.cargo/config.toml` files

## Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

This RFC adds a few new configuration options to [cargo configuration](https://doc.rust-lang.org/cargo/reference/config.html).

### Added to [Configuration Format](https://doc.rust-lang.org/cargo/reference/config.html#configuration-format)

```toml
[resolver]
incompatible-publish-age = "deny" # Specifies how resolver reacts to these

[registries.<name>]
min-publish-age = "..."  # Override `registry.global-min-publish-age` for this registry

[registry]

> min-publish-age = "0"  # Override `registry.global-min-publish-age` for crates.io

Reviewer 5: Should this use `"..."` since it isn't the default?
```suggestion
min-publish-age = "..."  # Override `registry.global-min-publish-age` for crates.io
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..fcbe967ea774398cb47185f7aed66e48905ccb2a)*

Reviewer: Fixed in 75ce9692cb0377f6a320e8787d4f3ef7412dded6

global-min-publish-age = "0"  # Minimum time span allowed for packages from this registry
 ```

### Added to [`[resolver]`](https://doc.rust-lang.org/cargo/reference/config.html#resolver)

#### `resolver.incompatible-publish-age`

* Type: String
* Default: `"deny"`
* Environment: `CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE`

When resolving the version of a dependency, specify the behavior for versions with a `pubtime` (if present) that is incompatible with `registry.min-publish-age`. Values include:

* `allow`: treat pubtime-incompatible versions like any other version

> * `deny`: treat pubtime-incompatible versions like yanked versions

Reviewer 5: In the docs, can we describe the behavior independent of yanked?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..fcbe967ea774398cb47185f7aed66e48905ccb2a)*

Reviewer: Fixed in 3f26931afe44ecfea4ff0c5595e26c22c55c771d

  * For [source replacement],
    registry mirrors are expected to preserve `pubtime` from the index to be applicable.
    Local registries and directory (vendored) sources typically don't have `pubtime`,
    so the check does not apply.
* When resolving dependencies,
  pubtime-incompatible versions follow the same rules as [yanked versions]:
  * They are not considered for new resolves,
    but if already present in `Cargo.lock`, they are left in place.
  * If no version satisfies both the version requirement and the minimum publish age,
    the resolve will error.
  * A status message will be printed when selecting a non-latest version
    as well for incompatible versions.
  * [`cargo update --precise`] can override the policy,
    just as it can select a yanked version.

[source replacement]: https://doc.rust-lang.org/cargo/reference/source-replacement.html
[yanked versions]: https://doc.rust-lang.org/cargo/reference/resolver.html#yanked-versions
[`cargo update --precise`]: https://doc.rust-lang.org/cargo/commands/cargo-update.html#option-cargo-update---precise

## Drawbacks

> [drawbacks]: #drawbacks

Reviewer 5: The `[resolver]` configs support a disjoint set of values which could be confusing

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..fcbe967ea774398cb47185f7aed66e48905ccb2a)*

Reviewer: Fixed in 51c71b50343b521b8ef67a155de45cddf79d3c84

clear that this only applies to crates that are published in a registry.

`publish` is redundant with this being in the `registry` table.
This helps with the above disambiguation and for clarity in discussing this as a shorthand.

`cooldown` was avoided due to term generally referring to throttling while we are looking for a certain maturity.

### Starting with `deny`

`resolver.incompatible-publish-age` starts with `allow` and `deny`.

Unlike `resolver.incompatible-rust-versions` which starts with `fallback`,
`deny` is viable here because `pubtime` data is exhaustive.
crates.io sets it for every version once backfilled,
so there are no gaps that would cause spurious errors.

A `fallback` option would deprioritize too-new versions but still allow them as a last resort.
This is deferred because it opens the yank attack vector:
a malicious actor with right permissions could publish a malicious version and yank the safe versions.
It then forces the resolver to fall back to the malicious too-new version.
`deny` prevents this by erroring instead of falling back.

`fallback` may be useful in the future for risk-tolerant workflows
that prefer a degraded resolve over an error,

> particularly when combined with other tools that validate pubtime-incompatible versions.

Reviewer 5: `fallback` could also be useful for the `CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE=fallback cargo update clap --precise 4.5.3` case as it ensures you get the bare minimum of new package versions

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..fcbe967ea774398cb47185f7aed66e48905ccb2a)*

Reviewer: Fixed in  0d167e7b289311477c54d065a2e9b7d50e48a5e6

* Would it be better to have `registry.min-publish-age` be the global setting, and `registries.crates-io.min-publish-age` be the setting for the crates.io registry?
  The current proposal is based on precedent of "credential-provider" and "global-credential-provider", but perhaps we shouldn't follow that precedent?
* How do we make it clear when things are held back?
  * The "locking" message for [Cargo time machine (generate lock files based on old registry state) #5221](https://github.com/rust-lang/cargo/issues/5221) lists one time but the time here is dependent on where any given package is from
  * We list MSRVs for unselected packages, should we also list publish times? I'm assuming that should be in local time
  * Locking message for [Cargo time machine (generate lock files based on old registry state) #5221](https://github.com/rust-lang/cargo/issues/5221) is in UTC time, see [Tracking Issue for _lockfile-publish-time_ #16271](https://github.com/rust-lang/cargo/issues/16271), when relative time differences likely make local time more relevant
* Implementation wise, will there be much complexity in getting per registry information into `VersionPreferences` and using it?

> * Is `resolver.incompatible-publish-age` needed at all?

Reviewer 5: I think there is enough of a case to not have this as an unresolved question, see https://github.com/rust-lang/rfcs/pull/3923#discussion_r3154604148

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..fcbe967ea774398cb47185f7aed66e48905ccb2a)*

Reviewer: Removed in  a2a748bdef618dfd9fb2725192212b224843907b

  - run: rustup update stable && rustup default stable
  - run: cargo update --verbose
  - run: cargo build --verbose
  - run: cargo test --verbose
```

This will mean that:

- Locally, `cargo update` will only select versions older than the minimum publish age,
  e.g., `some-package@1.2.3`
- This CI job will verify the latest versions of your dependencies,
  e.g., `some-package@1.3.0`

### Per-registry configuration

It is also possible to configure the `min-publish-age` per cargo registry.
`registries.<name>.min-publish-age` sets the minimum publish age for the `<name>` registry.
And `registry.min-publish-age` sets it for crates.io.

For example:
```toml
[registries.my-org]
index = "https://my.org"
min-publish-age = "0" # this registry is fully trusted

[registry]
# Default for any registry without a specific value
global-min-publish-age = "14 days"
# Value to use for crates.io
min-publish-age = "5 days"
```

This will use a minimum publish age of
- 5 days for crates.io
- no minimum for `my-org`
- 14 days for any other registry.

### When no version matches

If no version of a dependency satisfies both the version requirement and the minimum publish age,

> the resolve will error, similar to when all matching versions are yanked:
> 
> ```console
> $ cargo update
> error: failed to select a version for the requirement `some-package = "^1.3"`
>   version 1.3.0 is too new (published 2 days ago, minimum age 14 days)
> ```

Reviewer 5: More as a reminder to ourselves to see if we can find ways to help users
```suggestion
```console
$ cargo update
error: failed to select a version for the requirement `some-package = "^1.3"`
  version 1.3.0 is too new (published 2 days ago, minimum age 14 days)
help: to preserve the min-publish-age, downgrade the version requirement to `"1.1"`
help: to use `"1.3"` anyways, re-resolve with `CARGO_RESOLVER_INCOMPATIBLE_PUBLISH_AGE=allow`
```
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..ac18038da098ade6a0581ad99595769e7707601f)*

Reviewer: Expanded in 6f86e397fb9c9a0eb1cdbe17edb2da4a273d272f

the resolver will ignore these versions
unless they already exist in the `Cargo.lock` file.
Setting the config to `allow` would disable the check,
which if combined with `cargo update --precise`,
cargo would pull in a specific version and its transitive dependencies.

[`resolver.incompatible-publish-age`]: config.md#resolverincompatible-publish-age
[`min-publish-age`]: config.md#registryglobal-min-publish-age
[Yanked versions]: https://doc.rust-lang.org/cargo/reference/resolver.html#yanked-versions

### Applicability

The minimum publish age check applies to the following dependency sources:

* Dependencies fetched from a registry that publishes `pubtime`, such as crates.io.
* Does not apply to git or path dependencies,
  in part because there is not always an obvious publish time,
  or a way to find alternative versions.
* Does not apply to registries that don't set `pubtime`,
  as there is no reliable way to know when the version was published.
* For [source replacement],
  registry mirrors are expected to preserve `pubtime` from the index to be applicable.
  Local registries and directory (vendored) sources typically don't have `pubtime`,
  so the check does not apply.

> * `cargo install` also skips pubtime-incompatible versions.

Reviewer 5: We need a rationale listed for this

```suggestion
* `cargo install` also skips pubtime-incompatible versions as the `resolver` table is documented as not applying to `cargo install`
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..ac18038da098ade6a0581ad99595769e7707601f)*

Reviewer: Clarified in f1244873dce8ced408286231d87bcbaa7cb98cc9

[`[registry]`](https://doc.rust-lang.org/cargo/reference/config.html#registry)

> * Set default registry
> * Sets credential providers for all registries
> * Sets crates.io values

[`[registries]`](https://doc.rust-lang.org/cargo/reference/config.html#registries)

> * Sets registry specific values

`yanked`: can't do new resolves to it but left in if already there. `--precise` can force it but that doesn't apply recursively.

pre-release: requires opt-in through version requirement. Unstable support to force it with `--precise` but that doesn't apply recursively.

## Unresolved Questions
[unresolved-questions]: #unresolved-questions

### Before RFC acceptance

* Can we, and should we, make any guarantees about security when using this feature,
  such as "a malicious version of a crate will not compromise the build
  if published within the minimum publish age window"?

### Before stabilization

* Would it be better to have `registry.min-publish-age` be the global setting, and `registries.crates-io.min-publish-age` be the setting for the crates.io registry?

>   The current proposal is based on precedent of "credential-provider" and "global-credential-provider", but perhaps we shouldn't follow that precedent?

Reviewer 5: Should we put a stake in the ground and remove this?  We have the rationale given.  Team members can raise concerns.  We can still re-evaluate as we go if we find a reason to.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..ac18038da098ade6a0581ad99595769e7707601f)*

Reviewer: Removed in 8e5538d3e43f2ebf9b35a39728559259a1f1d46c

the resolver will ignore these versions
unless they already exist in the `Cargo.lock` file.
Setting the config to `allow` would disable the check,
which if combined with `cargo update --precise`,
cargo would pull in a specific version and its transitive dependencies.

[`resolver.incompatible-publish-age`]: config.md#resolverincompatible-publish-age
[`min-publish-age`]: config.md#registryglobal-min-publish-age
[Yanked versions]: https://doc.rust-lang.org/cargo/reference/resolver.html#yanked-versions

### Applicability

The minimum publish age check applies to the following dependency sources:

* Dependencies fetched from a registry that publishes `pubtime`, such as crates.io.
* Does not apply to git or path dependencies,
  in part because there is not always an obvious publish time,
  or a way to find alternative versions.
* Does not apply to registries that don't set `pubtime`,
  as there is no reliable way to know when the version was published.
* For [source replacement],
  registry mirrors are expected to preserve `pubtime` from the index to be applicable.
  Local registries and directory (vendored) sources typically don't have `pubtime`,
  so the check does not apply.

> * `cargo install` also skips pubtime-incompatible versions

Reviewer 10: What do you mean by "cargo install also skips pubtime-incompatible versions"? Did we already support this feature? I got confused. Or did I overlook anything?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..f1244873dce8ced408286231d87bcbaa7cb98cc9)*

Reviewer: That is under Applicability section, and it proposes that pubtime-incompatible config options in `[resolver]` don't affect any `cargo install` behavior at this moment. The "also" refers to just like other resolver options.

Reviewer: Disambiguated in https://github.com/rust-lang/rfcs/pull/3923/changes/6dca5a1de5e4bee7f961b51dd88b617ab42dc22b

  `resolver.incompatible-publish-age` cannot override this as the `[resolver]` table does not apply to `cargo install`.

[source replacement]: https://doc.rust-lang.org/cargo/reference/source-replacement.html

## Drawbacks
[drawbacks]: #drawbacks

### Slower problem discovery

> The biggest drawback is that if this is widely used, it could potentially lead to it taking longer for problems to be discovered after a version is published.

Reviewer 11: To add to this, discovering an issue later may make it harder to address.

Right now if a problem in a new crate release is discovered right away then the version can be yanked and the change reverted. This can be helpful in the case of an accidental semver break in a minor version or a newly added API with a serious flaw. However, there's only a narrow (and ambiguous) window of time before it is too late to yank and revert. Eventually users are going to start relying on the new behavior.

With perfect visibility into the speed of version rollouts, crate authors might be able to account for the difference. But I fear that "someone's CI detected the breakage 2 hours after release" is much more intuitively OK to yank for than "its been 5 days but I bet most people use a 7+ day cooldown"

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3923/c143e315774f746d667c5eecd95a8ed999e8a729..6dca5a1de5e4bee7f961b51dd88b617ab42dc22b)*

Reviewer: Good point. I've added a paragraph acknowledging this: https://github.com/rust-lang/rfcs/pull/3923/commits/2561c143ecfb29e6c6ba595303f3591e04ce1030

Reviewer 12: Given the supply chain attacks currently hitting NPM I think there's really no alternative to adding this.
